torch/utils/_content_store.py

```diff
@@ -0,0 +1,156 @@
+# users.
+
+import os.path
+import struct
+from typing import Optional
+
+import torch
+import torch._prims as prims
+import torch._utils
+import torch.nn.functional as F
+
+from torch.multiprocessing.reductions import StorageWeakRef
+
+
+def hash_storage(storage):
+    # Use of torch.compile is mandatory for (1) good memory usage
+    # and (2) xor_sum implementation
+    @torch.compile(dynamic=True)
+    def kernel(x):
+        # The randint calls are carefully written to hit things we
+        # have lowerings for in inductor.  Lack of unsigned 32-bit integer
+        # is a pain.
+        a = torch.randint(
+            -(2**31), 2**31, x.shape, device=x.device, dtype=torch.int32
+        ).abs()
+        a = ((a % (2**31 - 1)) + 1).long()
+        b = (
+            torch.randint(
+                -(2**31), 2**31, x.shape, device=x.device, dtype=torch.int32
+            )
+            .abs()
+            .long()
+        )
+        # This is a standard shift-multiply universal hash family
+        # plus xor sum hash, using Philox to generate random numbers.
+        # Our Philox RNG is not deterministic across devices so
+        # don't use this for stable hashing.
+        #
+        # This assumes fixed length so you're also obligated to bucket
+        # by the length of tensor as well
+        return prims.xor_sum((a * x + b).int(), [0])
+
+    with torch.random.fork_rng(
+        [storage.device] if storage.device.type != "cpu" else []
+    ):
+        torch.manual_seed(0)  # this can be anything, just needs to be fixed
+        x = torch.empty(0, dtype=torch.uint8, device=storage.device).set_(storage)
+        # The dtype-casting view cannot be compiled, and so the
+        # padding/reshaping also needs to be done externally even
+        # though it could be profitably fused
+        pad = x.numel() % 4
+        if pad > 0:
+            x = F.pad(x, (0, pad), "constant", 0)
+        x = x.view(torch.int32)
+        torch._dynamo.mark_dynamic(x, 0)
+        # We run the 32-bit hash five times with differing parameters to
+        # reduce chance of collision
+        ITER = 5
+        cs = [kernel(x).item() for _ in range(ITER)]
+        return struct.pack(">" + "i" * ITER, *cs).hex()
+
+
+class ContentStoreWriter:
+    # Structure:
+    #   storages/
+    #     00/
+    #       0000..00
+    #   tensors/
+    #     name
+    def __init__(self, loc):
+        self.loc = loc
+        self.seen_storage_hashes = set()
+
+    def write_storage(self, storage):
+        h = hash_storage(storage)
+        if h in self.seen_storage_hashes:
+            return h
+        # TODO: consider not using torch.save for this; we don't actually
+        # need any metadata for the storage
+        subfolder = os.path.join(self.loc, "storages", h[:2])
+        os.makedirs(subfolder, exist_ok=True)
+        torch.save(storage, os.path.join(subfolder, h))
+        self.seen_storage_hashes.add(h)
```

Comment:
nit: If you write a helper, you can use @lru_cache for this and not need to manage your own structure

Reply:
Can't easily use lru_cache as storage isn't hashable. And it seems dorky to pass the StorageWeakRef as the input to the function?